Add Stryker4s sbt performance blog (#292)
stryker-mutator/stryker-mutator.github.io: blog/2021-06-14-stryker4s-sbt-performance.mdx

Imagine you're on your way to a sunny holiday destination. You've checked in your luggage, gone through customs, and boarded the plane. Then, the plane moves forward a couple of meters, and you're thrown out and sent back to customs before going through the whole thing again. It'll take years before you get to sip on your umbrella cocktail 🍹! This is exactly how Stryker4s has been running your code… until now.

## 📃 In short

A lot has changed in recent versions of Stryker4s. Stryker4s now runs your tests in an entirely different way, bringing big performance improvements. As an example, running Stryker4s on the Stryker4s codebase used to take about 40 minutes on `0.9.1`. On`0.12.1`, this has been reduced to a fast 40 seconds. A short list of some new features:

- ✨ New testrunner boasting improved performance

> - 🔀 Concurrency for even faster performance

Reviewer: Adjectives like "even" (faster), "a lot", "very", etc often have the opposite effect on the reader. They might think: "wait a second, is it _really_ faster? Wasn't it fast before?". 

Suggestion: "Concurrency to improve speed by using all available cores"

- 🕵️‍♀️ Coverage analysis to skip mutants without coverage
- ⏱️ Timeout detection in tests
- 👽 New mutator: regular expressions. Thanks to [Weapon regeX](https://github.com/stryker-mutator/weapon-regex/)!

In the rest of this blog post, we'll do a deep-dive into the increased performance. Let's start by taking a look at the older version of Stryker4s.

## 🐌 The old

In the older sbt plugin, there was a lot of overhead in running your tests. As a quick reminder, Stryker4s uses [mutation switching](https://stryker-mutator.io/blog/mutation-switching/), so it only has to compile your code once, instead of for each mutant. After mutating your source code, Stryker4s would essentially just call `sbt test` with the mutated code for every mutant. This means sbt would:

> 1. look for code changes (which there never are).

Reviewer: Mixed casing in this ordered list.

1. make sure all compiled code is up-to-date (which it always is).
1. Start a new Java process with your class files, dependencies, and a small sbt entry point library.
1. Run your tests and report the results back to the main sbt process.
1. Repeat for the next mutant.

import Mermaid from '@theme/Mermaid';

<Mermaid
  _key="sbt-old"
  chart={`

> sequenceDiagram

Reviewer: Suggestion:

```
sequenceDiagram
  participant Stryker4s
  participant sbt
  participant sbt test process
  Stryker4s->>+sbt: Compile mutated code
  sbt-->>-Stryker4s: bytecode
  loop For each mutant
    Stryker4s->>+sbt: executeTests
    sbt->>sbt: Check for changes?
    sbt->>+sbt test process: Start new process and run tests
    sbt test process-->>-sbt: Report test results
    sbt-->>-Stryker4s: Tests successful Y/N
  end
```

(sorry, don't know how to do multiline suggestions 🙈)

Author: This is how I had it first. It makes sense for this image, but then sbt and Stryker4s are swapped in the diagram below. What do you think is best?

  participant sbt
  participant Stryker4s
  participant sbt test process
  Stryker4s->>sbt: Compile mutated code
  loop For each mutant
    Stryker4s->>sbt: executeTests
    sbt->>sbt: Check for changes?
    sbt->>+sbt test process: Start new process and run tests
    sbt test process->>-sbt: Report test results
    sbt->>Stryker4s: Tests successful Y/N
  end
`}
/>

> Doing those checks and starting a new process make a lot of sense for a development workflow, but it's a lot of overhead for mutation testing. This came from the original implementation of the sbt plugin and hasn't changed much in the two years that the plugin has existed. Time for some improvements!

Reviewer: ```suggestion
Doing those checks and starting a new process makes a lot of sense for a development workflow, but not for mutation testing. This implementation of the test runner hasn't changed much in the two years that the plugin has existed. Time for some improvements!
```

## 🐣 The new

It's clear the process can be optimized. Once Stryker4s is set up and all mutations have been compiled, the only thing we really want to do is run some tests and get a result. Sbt doesn't give plugins full control over the entire test process, but it does have all the information about _what_ your project looks like. This is something we can work with!

### 🏗️ Meet: build info

> The first thing we need to do is figure out what tests your project contains. One of the great things about sbt is its standardization and extensibility as a build tool. Because sbt knows exactly what your project looks like, plugins can retrieve all this information. To find out what tests you have, Stryker4s uses three pieces of information from sbt:

Reviewer: Maybe instead of "we" and "your" we can use "Stryker4s" and "our"?

I.e.: The first thing Stryker4s needs to do is figure out what tests our project contains.

> 1. `fullClasspath` location of all dependency JARs (including test-dependencies) and `.class` files of your compiled code.

Reviewer: Maybe we can link to the official docs here?

Author: Not all of these have relevant documentation unfortunately 😢

You can try running these yourself in a sbt console to see how Stryker4s looks at your project:

```
sbt:stryker4s-core> show fullClasspath
[info] * Attributed(~/stryker4s/core/target/jvm-2.13/classes)
[info] * Attributed(~/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-
library/2.13.4/scala-library-2.13.4.jar)
[info] * Attributed(~/.cache/coursier/v1/https/repo1.maven.org/maven2/com/github/pathikrit/
better-files_2.13/3.9.1/better-files_2.13-3.9.1.jar)
...
sbt:stryker4s-core> show loadedTestFrameworks
[info] Map(TestFramework(org.scalatest.tools.Framework, org.scalatest.tools.ScalaTestFramework) -> org
.scalatest.tools.Framework@2b9d52d5)
...
sbt:stryker4s-core> show testGrouping
[info] * (<default>,Vector(Test stryker4s.config.TestFilterTest : subclass(false, org.scalatest.Suite)
, Test stryker4s.extension.FileExtensionsTest : subclass(false, org.scalatest.Suite), Test stryker4s.m
utants.MutatorTest : subclass(false, org.scalatest.Suite), Test stryker4s.mutants.AddAllMutationsTest
: subclass(false, org.scalatest.Suite), ...
```

We now know what tests are in your project. Let's have a look at running those tests:

### 🧪 Meet: test interfaces

> This is something our friends in the NodeJS world are (rightfully) a little [jealous of](https://stryker-mutator.io/docs/stryker-js/plugins#test-runners). When you call `sbt test`, sbt will look for any registered test frameworks that implement the [test-interface](https://github.com/sbt/test-interface) (from `loadedTestFrameworks`). This is a small collection of interfaces that let a test framework tell sbt how to run tests in that framework. Because all test frameworks in sbt implement this interface, we have a single programmatic way to call any test framework. It doesn't matter if you're using ScalaTest, MUnit, Weaver-test or anything else. Using a bit of reflection, we can create an instance of the test framework runner and run tests:

Reviewer: ```suggestion
This is something our friends in the NodeJS world are (rightfully) a little [jealous of](https://stryker-mutator.io/docs/stryker-js/plugins#test-runners) 😏. When you call `sbt test`, sbt will look for any registered test frameworks that implement the [test-interface](https://github.com/sbt/test-interface) (from `loadedTestFrameworks`). This is a small collection of interfaces that let a test framework tell sbt how to run tests in that framework. Because all test frameworks in sbt implement this interface, we have a single programmatic way to call any test framework. It doesn't matter if you're using ScalaTest, MUnit, Weaver-test or anything else. Using a bit of reflection, we can create an instance of the test framework runner and run tests:
```

```scala
val framework = classLoader.loadClass(testFrameworkClass).getConstructor().newInstance().asInstanceOf[Framework]
val runner = framework.runner(args, remoteArgs, classLoader)

val testTasks: Array[Task] = runner.tasks(testGrouping.taskDefs)

testTasks.foreach(testTask => {
  testTask.execute()
})
```

### 🔀 Meet: concurrency

> When running your unit tests, Stryker4s activates each mutant separately and runs your unit tests against _only_ the change that mutant creates. It won't activate two mutants at the same time because there's no way to know which of the two mutants gave a result. However, there's no rule that says we can't start multiple Java processes at the same time and run those in parallel. Each process has its own separate memory, threads, and active mutant. This way, if one mutant result in an `OutOfMemoryError` it won't impact the others. Starting multiple testrunner workers also improves performance by making better use of multi-core CPUs.

Reviewer: > It won't activate two mutants at the same time because there's no way to know which of the two mutants gave a result

Well... there is. In fact, Stryker.NET does exactly that. You just need to make sure the mutants don't have overlapping tests.

Author: Yeah that crossed my mind too. But when you have per-test coverage (which is possible with this approach) it doesn't make sense

Reviewer: Sure, but why mention it at all? It's not needed for the concept of concurrency

### 🧰 Putting it all together

Stryker4s can start a new process with all the information collected above. Because Stryker4s now fully controls how it runs your tests, it can eliminate any overhead of starting new processes or looking for source code changes. It looks a little something like this:

1. Start new worker processes in parallel with the full test classpath and send them all the needed test setup.
1. Create a new instance of the test framework in the created process.
1. For each mutant, programmatically run the tests on the worker process.

<Mermaid
  _key="sbt-new"
  chart={`

> sequenceDiagram

Reviewer: Suggestion:

```
sequenceDiagram
  participant Stryker4s
  participant sbt
  participant testrunner worker
  Stryker4s->>+sbt: compile and retrieve build info
  sbt-->>-Stryker4s: return project info
  Stryker4s->>+testrunner worker: start <n> new workers
  loop For each mutant
    Stryker4s->>testrunner worker: run tests
    testrunner worker-->>Stryker4s: report test result
  end
  Stryker4s->>testrunner worker: close test worker
  testrunner worker-->>-Stryker4s: 
```

Author: I'm confused now. In this diagram the suggestion is to have Stryker4s first and then sbt, but in the first one to have sbt first and then Stryker4s?

Reviewer: Both of them should have Stryker4s first, but I didn't try them locally, just in the mermaidjs playground
